### PR TITLE
Support DDP for BatchSizeFinder

### LINKED ISSUE #4424: Batch size finder fails in DDP mode
## 🐛 Bug

The batch size finder fails in DDP mode, as `os.remove` is [called](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/tuner/batch_size_scaling.py#L113) from each process trying to load and delete the same model checkpoint.
The batch size finder works if the `os.remove` line is commented out (leaving the temporary checkpoint).

### To Reproduce

Run` trainer.tune()` in DDP mode.

### Environment

* CUDA:
	- GPU:
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.3
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0
	- pytorch-lightning: 1.0.4
	- tqdm:              4.51.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.6.9
	- version:           #25~18.04.1-Ubuntu SMP Fri Jun 5 15:18:30 UTC 2020

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #4424

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @akihironitta

### CODE REVIEW

**Review by akihironitta — APPROVED** [review]

**Comment by awaelchli** [comment]
After studying this a bit, I realized that there is a greater challenge here we need to solve. The batch size finder algorithm finishes the trials successfully when a OOM error is raised. However, there can be very small differences in how memory is used on individual the GPUs. Furthermore, one might train on similar GPUs that have different amounts of memory. 
These edge cases can cause the conditions on each rank to be different, and the BS finder algorithm can exit earlier on some ranks than others. This is problematic, because if the model runs in DDP, it has to be synchronous. 

In order to do this properly, we need to synchronize all the conditions and only exit when all ranks can exit / switch code path the same way. This has to be applied in places like here:
https://github.com/Lightning-AI/lightning/blob/f1bd68da3b6e1163c5769928fd57740feb5bbcc0/src/pytorch_lightning/tuner/batch_size_scaling.py#L159-L179

A refactor of this try-except block is needed for a structure that allows to synchronize the stopping condition over all ranks.
